### PR TITLE
Add SSL support

### LINKED ISSUE #15: Support SSL

### PR DESCRIPTION
**Pull request by nonnontrivial**
Resolves #15

### CODE REVIEW

**Comment by bartlomieju** [comment]
@nonnontrivial good start; I might be slow to review this week though. One nit pick; integration test is a must in this case; so we're gonna need a self signed certificate for that. You can take example from [`deno`](https://github.com/denoland/deno/blob/master/cli/tests/tls/README.md) on how to handle that and add tests

**Comment by nonnontrivial** [comment]
@bartlomieju sure. Two questions.

1. Is this to say that tests will never pass unless the machine running the tests already has a valid certificate in some expected location?
2. Are the tests expected to pass for `deno 0.34.0`? Also, how should `tsconfig.test.json` be defined? This file is referenced in the docs but untracked.

**Comment by bartlomieju** [comment]
> @bartlomieju sure. Two questions.
> 
> 1. Is this to say that tests will never pass unless the machine running the tests already has a valid certificate in some expected location?

Yeah, you can put it into `tests/cert/` directory.

> 2. Are the tests expected to pass for `deno 0.34.0`? Also, how should `tsconfig.test.json` be defined? This file is referenced in the docs but untracked.

Sure. I think I need to update documentation 😅 just ignore `tsconfig.test.json` bit

**Comment by jakajancar** [comment]
Hey, what is still missing here? Need help?

**Comment by nonnontrivial** [comment]
Hey @jakajancar . A few things
- [ ] `--allow-read` flag seemingly needs to be provided to the CI
- [ ] Clarity on what needs to happen aside from setting `conn` on the `Connection` instance to `Deno.connectTLS`
- [ ] Clarity on what other tests should be included

**Comment by jakajancar** [comment]
For #1, can it not be just added [here](https://github.com/buildondata/deno-postgres/blob/master/.travis.yml#L17)?

For #3, TravisCI's postgres supports SSL, so you if you change [this](https://github.com/buildondata/deno-postgres/blob/master/tests/constants.ts#L14-L20) to use it, I would imagine that would do it.

**Comment by halvardssm** [comment]
Hi @nonnontrivial! Are you still working on this issue? Do you need any help?

**Comment by nonnontrivial** [comment]
Hey @halvardssm . Thank you for reaching out.

Currently, the CI fails on this PR with `deno fmt --check`. Unsure why this would be.

Aside from this, two things I'm unclear on:

- How else does `this.conn` of `Connection` class need to be handled in the case that it is `Deno.connectTLS`?
- What tests are lacking?

**Comment by halvardssm** [comment]
@nonnontrivial you can pull from master since the ci was changed form travis to github, that should hopefully fix the issue of `deno fmt --check` failing. If that doesnt work, you can run `deno fmt` locally, that should clean up the formatting. 

For `this.conn` I am not entirelly sure. I can take a deeper look into the code to see if I figure it out.

Regarding tests, it seems like you already have some cert files. Test that the connection works, and that you can select/insert using the tls connection. You can look at `tests/client.ts` and maybe add a secure connection test there? The most important with the tests are that they cover most cases with the connection e.g. what should happen if a connection succeeds, breaks, or postgres refuses the connection for some reason.

**Comment by halvardssm** [comment]
@nonnontrivial From what I can see, you dont have to change anything with `this.conn`, did you come across anything which suggest that you have to change it? If you point me at the issue, I can maybe help you out with it.

**Comment by nonnontrivial** [comment]
Thank you, @halvardssm . Ran `deno fmt` (with `deno 0.35.0`). CI still does not like this. Running the exact command found in actions on my machine results in no error.

Adding tests you recommended in your previous comment.

**Comment by halvardssm** [comment]
@nonnontrivial Thats interresting 🤔 On my machine it shows up, can it be something regarding your deno version? We are at 0.41.0 at this moment, so maybe run `deno upgrade`? 
Screenshot 2020-04-25 at 16 27 02

File: connection.ts [hunk 1]
```diff
@@ -134,12 +134,22 @@ export class Connection {
...
-    const { host, port } = this.connParams;
-    this.conn = await Deno.connect({
-      port: parseInt(port, 10),
-      hostname: host,
-    });
+  public async startup() {
+    const { host, port, cert_file } = this.connParams;
+    const parsedPort = parseInt(port, 10);
+
+    if (typeof cert_file !== "undefined") {
```
Comment (halvardssm) [inline connection.ts:141]: Can you add a test to what should happen if cert_file is not found locally, and if it is not valid?

Comment (nonnontrivial) [reply connection.ts:141]: Okay. I'm assuming we want this to throw in the case that it's invalid?

Comment (halvardssm) [reply connection.ts:141]: Yeah, that might be the best. Just make sure to pass a relevant throw message 👍

File: connection_params.ts [hunk 2]
```diff
@@ -50,6 +50,7 @@ export interface IConnectionParams {
   user?: string;
   password?: string;
   application_name?: string;
+  cert_file?: string;
```
Comment (bartlomieju) [inline connection_params.ts:53]: I'm not sure why I used `snake_case` here :| we should probably use `camelCase` for consistency in code style

File: connection.ts [hunk 3]
```diff
@@ -134,12 +134,22 @@ export class Connection {
...
+      this.conn = await Deno.connectTLS({
+        port: parsedPort,
+        hostname: host,
+        certFile: cert_file,
```
Comment (bartlomieju) [inline connection.ts:145]: 👍

**Comment by bartlomieju** [comment]
@nonnontrivial sorry for slow review; the PR looks good; if you could only change `cert_name` and `application_name` to be camel case I'm ok to land this change

**Comment by hayd** [comment]
Would be great to see this green/rebased!

**Comment by hayd** [comment]
I tried a squash and a rebase here https://github.com/hayd/deno-postgres/commit/4a93c0adcd44ec60cc3850a080cf343a51f18982 (that's passing the tests).

My worry is that the test is pretty weak... I wonder if there is some way to actually test something (e.g. that it's used connectTls rather than connect). 🤔

**Comment by bartlomieju** [comment]
> I tried a squash and a rebase here [hayd@4a93c0a](https://github.com/hayd/deno-postgres/commit/4a93c0adcd44ec60cc3850a080cf343a51f18982) (that's passing the tests).
> 
> My worry is that the test is pretty weak... I wonder if there is some way to actually test something (e.g. that it's used connectTls rather than connect). 🤔

It should be possible to configure 2 separate Postgres instances in CI: one should be HTTP only; the other HTTPS only.

**Comment by hayd** [comment]
I'm not sure the incantation to do this in the github action, if anyone has any ideas that would be very helpful (do you pass additional options with the cert etc.?):

https://github.com/deno-postgres/deno-postgres/blob/master/.github/workflows/ci.yml

**Comment by halvardssm** [comment]
@hayd You can add a new service in the `ci.yml` by duplicating the existing postgres service and name it `postgres_sql` or something similar and change the port. To enable ssl you can follow this thread (i think) https://gist.github.com/mrw34/c97bb03ea1054afb551886ffc8b63c3b. Let me know if anything is still unclear, or if you stumble across any issues 💪

**Comment by halvardssm** [comment]
@hayd adding 
```
volumes:
      - path_to_cert:/var/lib/postgresql/server.crt
      - path_to_key:/var/lib/postgresql/server.key
```

and adding `--entrypoint` to options to include 
```
-c ssl=on -c ssl_cert_file=/var/lib/postgresql/server.crt -c ssl_key_file=/var/lib/postgresql/server.key
```
might be a solution. It's possible you can omit `-c ssl_key_file` and `-c ssl_cert_file` but I havent tested it.

**Comment by halvardssm** [comment]
@hayd (sorry to bug you so much) I think, according to the [docs](https://www.postgresql.org/docs/9.1/ssl-tcp.html) that you don't even have to change the entrypoint as long as the cert and key are available on boot, so by adding the volumes, it should work. 
> To start in SSL mode, the files server.crt and server.key must exist in the server's data directory.

**Comment by jakajancar** [comment]
I don't think this implementation will work. You need to use `Deno.startTls()` because PostgreSQL expects a magic number before beginning TLS negotiation. I just published a more full-featured PostgreSQL client, you can see how it can be done here:

https://github.com/jakajancar/pgc4d

Specifically `connection.ts`.

**Comment by devanandp** [comment]
> I don't think this implementation will work. You need to use `Deno.startTls()` because PostgreSQL expects a magic number before beginning TLS negotiation. I just published a more full-featured PostgreSQL client, you can see how it can be done here:
> 
> https://github.com/jakajancar/pgc4d
> 
> Specifically `connection.ts`.

Can anyone tell why Iam getting this error while connecting to Heroku Postgres? Is this related to this issue?


WARN RS - rustls::session:718 - Sending fatal alert BadCertificate
error: Uncaught InvalidData: invalid certificate: UnknownIssuer

**Comment by nonnontrivial** [comment]
> I don't think this implementation will work. You need to use `Deno.startTls()` because PostgreSQL expects a magic number before beginning TLS negotiation. I just published a more full-featured PostgreSQL client, you can see how it can be done here:
> 
> https://github.com/jakajancar/pgc4d
> 
> Specifically `connection.ts`.

@bartlomieju can you confirm that this is true?

**Comment by OliverBrotchie** [comment]
@nonnontrivial @halvardssm Is this commit still being worked on?

**Comment by halvardssm** [comment]
I think @nonnontrivial is working on it but got stuck on the TSL connection part. Not sure what the current status is

**Comment by OliverBrotchie** [comment]
> I don't think this implementation will work. You need to use `Deno.startTls()` because PostgreSQL expects a magic number before beginning TLS negotiation. I just published a more full-featured PostgreSQL client, you can see how it can be done here:
> 
> https://github.com/jakajancar/pgc4d
> 
> Specifically `connection.ts`.
@jakajancar I'm getting 'Deno.startTls' is not a function when I run your example code.

@halvardssm is there any chance we could get basic SSL merged in? His last commit for this issue was two months ago and my team struggling to find an alternative to this.

**Comment by jakajancar** [comment]
> @jakajancar I'm getting 'Deno.startTls' is not a function when I run your example code

You need to run deno with --unstable for TLS.

**Comment by halvardssm** [comment]
@OliverBrotchie I am not sure, the current maintainer for this project is @hayd and the one who has been working on this until now has been @nonnontrivial , they can better help you out 👍

**Comment by nonnontrivial** [comment]
@OliverBrotchie there is uncertainty whether the approach taken in this PR will work (see https://github.com/deno-postgres/deno-postgres/pull/87#issuecomment-633294442) and how a test for it should be written (see https://github.com/deno-postgres/deno-postgres/pull/87#issuecomment-632995426).

**Comment by hayd** [comment]
Adding a test is the key, I think we just need to iterate on CI to get something, looking at pg4d may also help to see how it's done there . This is very important feature.

**Comment by ebebbington** [comment]
Wouldn't a successful connection and PING command be enough to test the TLS works? I guess under the hood, if the connection is there, Deno is correctly serving as TLS? More would be ideal of course and I guess this is more integration testing, but it's not like you can do a `fetch("https://redis_server")` for a web app that supports `serverTLS`

**Comment by ebebbington** [comment]
> @OliverBrotchie there is uncertainty whether the approach taken in this PR will work (see [#87 (comment)](https://github.com/deno-postgres/deno-postgres/pull/87#issuecomment-633294442)) and how a test for it should be written (see [#87 (comment)](https://github.com/deno-postgres/deno-postgres/pull/87#issuecomment-632995426)).

pgc4d managed to do it so it must eventually work right? I think to get it working:

* ssl must be setup on the postgres container (be it the ci container or user container)
* connect using `Deno.startTls`
* Have a dev cert (could use the one pgc4d is using?)

**Comment by Soremwar** [comment]
Closing in favor of #249

**Comment by borsemayur2** [comment]
> 
> 
> > I don't think this implementation will work. You need to use `Deno.startTls()` because PostgreSQL expects a magic number before beginning TLS negotiation. I just published a more full-featured PostgreSQL client, you can see how it can be done here:
> > https://github.com/jakajancar/pgc4d
> > Specifically `connection.ts`.
> 
> Can anyone tell why Iam getting this error while connecting to Heroku Postgres? Is this related to this issue?
> 
> WARN RS - rustls::session:718 - Sending fatal alert BadCertificate
> error: Uncaught InvalidData: invalid certificate: UnknownIssuer

I'm getting the same error. Were you able to use heroku postgres database?

**Pull request closed without merging**